Option to skip baselayout and creating /run and /var, and symlinking /var/run to /run

berney opened the pull request:
This happens after the last hook, finish_rootfs_build, so unless there's an option that allows it to be skipped, all images will always have `/run/` and `/var/` directories, and a symlink from `/var/run` to `/run`. I want this option to be able to skip this step and avoid creating the directories and symlink.

jbergstroem commented:
What processes/functionality/container actually needs `{/var}/run`? Perhaps just bundle the creation with that.

edannenberg commented on bob-core/build-root.sh:
```diff
@@ -367,7 +367,9 @@ generate_documentation_footer
 
 # /run symlink
-mkdir -p $EMERGE_ROOT/run $EMERGE_ROOT/var && ln -s /run $EMERGE_ROOT/var/run
+if [ -z "SKIP_VAR_RUN_SYMLINK" ]; then
```
Renaming this to `BOB_SKIP_VAR_RUN_SYMLINK` would enable setting the value once per namespace or even global. (All env starting with `BOB_` are exported from host to each build container)

berney replied:
Nice, I didn't realise that.

edannenberg commented on bob-core/build-root.sh:
```diff
@@ -367,7 +367,9 @@ generate_documentation_footer
 # /run symlink
-mkdir -p $EMERGE_ROOT/run $EMERGE_ROOT/var && ln -s /run $EMERGE_ROOT/var/run
+if [ -z "SKIP_VAR_RUN_SYMLINK" ]; then
+    mkdir -p $EMERGE_ROOT/run $EMERGE_ROOT/var && ln -s /run $EMERGE_ROOT/var/run
```
Cosmetics, but `$EMERGE_ROOT/{run,var}` is a bit nicer on the eyes.

edannenberg commented:
@jbergstroem Pretty sure `php-fpm` places it's pid there, generally all kinds of daemons may want to put their pid and similiar stuff there. Not sure it's a good idea to make it opt-in.

If there is a actual use case for using `PACKAGES` and not requiring `baselayout` we may want to bundle this into a `BUILD_MINIMAL` flag which would skip `baselayout`, run symlink etc. IIRC emerging packages at a custom root without `baselayout` was rather error prone.

berney commented:
I'll update this PR to incorporate your comments `BOB_SKIP_VAR_RUN_SYMLINKS` and `/{run,var}/`.

Should it be `BOB_BUILD_MINIMAL` or `BUILD_MINIMAL`? `BOB_BUILD_MINIMAL` right?

I was planning on a new PR to have  a `SKIP_BASELAYOUT`, so that `PACKAGES` could be used as normal but without automatically pulling in `baselayout` as well - and I was also thinking of a global flag like `BUILD_MINIMAL` (didn't have a name in mind) that would skip both the baselayout and this `/{run,var}/` and `/var/run`. Should I just add a `BOB_SKIP_BASELAYOUT` to this PR or separate PR?

jbergstroem commented:
> @edannenberg said:
> If there is a actual use case for using `PACKAGES` and not requiring baselayout we may want to bundle this into a `BUILD_MINIMAL` flag which would skip baselayout, run symlink etc. IIRC emerging packages at a custom root without baselayout was rather error prone.

This would be the proper way, but as you say I'm not sure either what we can expect in terms of fallout. Perhaps just give it a run and see how we end up.

edannenberg commented:
> I was planning on a new PR to have a SKIP_BASELAYOUT, so that PACKAGES could be used as normal but without automatically pulling in baselayout as well - and I was also thinking of a global flag like BUILD_MINIMAL (didn't have a name in mind) that would skip both the baselayout and this /{run,var}/ and /var/run. Should I just add a BOB_SKIP_BASELAYOUT to this PR or separate PR?

I think `BOB_SKIP_BASELAYOUT` is a very fitting, unless there is more you want to strip out that would make `BOB_BUILD_MINIMAL` maybe more fitting. Up to you. I'd say let's assume the baselayout should include the run symlink so just `BOB_SKIP_BASELAYOUT` for both is probably fine. And yea you can repurpose this PR.

> This would be the proper way, but as you say I'm not sure either what we can expect in terms of fallout. Perhaps just give it a run and see how we end up.

As it is not default I'm not too worried about fallout, people that use `BOB_SKIP_BASELAYOUT` are expected to know what they are doing. :stuck_out_tongue:

berney commented:
I updated the PR to have `BOB_SKIP_LAYOUT`, which skips both the emerging of baselayout and the making of the `/{var,run}` dirs and `/var/run` symlink.

I also rebased off `master`, and did a `git push --force-with-lease`, so the history has been re-written but GitHub seems happy, so it looks like I did things the right way.

I updated my figlet-musl and figlet-musl-static images to use this branch, and rebuilt the builder containers and my updated figlet-* images and it worked well. figlet-musl now uses `PACKAGES` to install figlet rather doing it manually and get the binary dynamically linked figlet without the automatic baselayout.

Manual testing and inspection of the new images looked good:
```
     _              _          _          _
 _ _(_)__ ___   _ _(_)__ ___  | |__  __ _| |__ _  _
| '_| / _/ -_) | '_| / _/ -_) | '_ \/ _` | '_ \ || |
|_| |_\__\___| |_| |_\__\___| |_.__/\__,_|_.__/\_, |
                                               |__/
```

edannenberg commented:
Excellent, squashed the commits, merged. Thanks!

edannenberg commented:
Test run with `BOB_SKIP_BASELAYOUT` enabled looked ok, this is now default behaviour and the variable got renamed to `BOB_INSTALL_BASELAYOUT`.
